File: worker/src/jams_worker/effort_score.py
"""Effort score normalization shared by worker scoring provider tests."""

from __future__ import annotations

import math
from typing import Any

Measure = dict[str, Any]

KIND_CATEGORY = {
    "clicks": "physical",
    "context_switch": "cognitive",
    "keypresses": "physical",
    "sentiment": "sentiment",
    "scrolls": "physical",
    # Words per minute is speech, not physical effort (preview review B6).
    "spoken_word": "speech",
    "time_segment": "time",
    "utterance": "speech",
}

# An utterance counts toward negative narration time at or below this value. Matches the report's
# "worth attention" line (apps/web/src/lib/report-moments.ts). It was -0.15 while the sentiment
# model was two-class and only produced extreme values; the three-class model leaves neutral
# lines slightly below zero, which -0.15 counted as negative. Keep in sync with effort-score.ts.
NEGATIVE_SENTIMENT_THRESHOLD = -0.3

SCORE_KIND_ORDER = [
    "clicks",
    "context_switch",
    "keypresses",
    "sentiment",
    "scrolls",
    "spoken_word",
    "time_segment",
    "utterance",
]

PER_MINUTE_SCALE = {
    "context_switch": 36.75,
    "clicks": 1,
    "keypresses": 1,
    "spoken_word": 2.47,
    "utterance": 10,
}


def round_whole(value: float) -> int:
    return math.floor(value + 0.5)


def round_tenth(value: float) -> float:
    return math.floor(value * 10 + 0.5) / 10


def clamp_score(value: float) -> float:
    return max(0, min(100, value))


def _measures_for_kind(measures: list[Measure], kind: str) -> list[Measure]:
    return [measure for measure in measures if measure.get("kind") == kind]


def _duration_minutes(video: dict[str, Any]) -> float:
    return float(video.get("duration_ms") or 0) / 60_000


def _per_minute_raw(kind: str, measures: list[Measure]) -> float:
    kind_measures = _measures_for_kind(measures, kind)
    if kind in {"keypresses", "spoken_word"}:
        return sum(float(measure.get("value_num") or 0) for measure in kind_measures)
    return float(len(kind_measures))


def _normalize_per_minute(
    kind: str,
    measures: list[Measure],
    video: dict[str, Any],
) -> dict[str, Any]:
    raw = _per_minute_raw(kind, measures)
    scale = PER_MINUTE_SCALE.get(kind, 1)
    minutes = _duration_minutes(video)
    rate = raw / minutes if minutes > 0 else 0
    return {
        "raw": int(raw) if raw.is_integer() else raw,
        "normalized": round_whole(clamp_score(rate * scale)),
    }


def _measure_end(measure: Measure) -> int:
    end_ms = measure.get("t_end_ms")
    return int(end_ms if end_ms is not None else measure["t_start_ms"])


def _normalize_negative_density(measures: list[Measure]) -> dict[str, Any]:
    sentiment_measures = _measures_for_kind(measures, "sentiment")
    narration_ms = sum(
        max(0, _measure_end(measure) - int(measure["t_start_ms"])) for measure in sentiment_measures
    )
    negative_ms = sum(
        max(0, _measure_end(measure) - int(measure["t_start_ms"]))
        for measure in sentiment_measures
        if float(measure.get("value_num") or 0) <= NEGATIVE_SENTIMENT_THRESHOLD
    )
    share = negative_ms / narration_ms if narration_ms > 0 else 0
    return {
        "raw": round_tenth(share),
        "normalized": round_whole(clamp_score(share * 250)),
    }


def _normalize_raw_minutes(video: dict[str, Any]) -> dict[str, Any]:
    minutes = _duration_minutes(video)
    return {
        "raw": round_tenth(minutes),
        "normalized": round_whole(clamp_score(minutes * (100 / 30))),
    }


def is_measured(kind: str, measures: list[Measure], video: dict[str, Any]) -> bool:
    """Whether this recording could measure the kind at all. Unmeasured is not zero.

    Speech kinds need audio (audio with silence is a real zero); sentiment needs speech to
    classify; context switches and time always come from the video; the experimental physical
    detectors count only when they produced something. Unknown audio is treated as present so
    older rows score as before.
    """
    if kind in {"spoken_word", "utterance"}:
        return video.get("has_audio") is not False
    if kind == "sentiment":
        return bool(_measures_for_kind(measures, "sentiment"))
    if kind in {"context_switch", "time_segment"}:
        return True
    return bool(_measures_for_kind(measures, kind))


def normalize(
    measures: list[Measure],
    video: dict[str, Any],
    normalization: dict[str, str],
) -> dict[str, dict[str, Any]]:
    scores: dict[str, dict[str, Any]] = {}
    for kind in SCORE_KIND_ORDER:
        mode = normalization.get(kind)
        if not mode:
            continue
        if mode == "per_minute":
            scores[kind] = {"kind": kind, **_normalize_per_minute(kind, measures, video)}
        elif mode == "neg_density":
            scores[kind] = {"kind": kind, **_normalize_negative_density(measures)}
        else:
            scores[kind] = {"kind": kind, **_normalize_raw_minutes(video)}
        scores[kind]["measured"] = is_measured(kind, measures, video)
    return scores


def score(normalized: dict[str, dict[str, Any]], weights: dict[str, float]) -> dict[str, Any]:
    weighted_kinds = [kind for kind in SCORE_KIND_ORDER if float(weights.get(kind) or 0) > 0]
    # Only kinds this recording could measure take part, in numerator and denominator alike, so a
    # missing signal never drags the score toward zero. A kind with no normalization is missing.
    active_kinds = [
        kind
        for kind in weighted_kinds
        if normalized.get(kind, {}).get("measured", kind in normalized)
    ]
    total_weight = sum(float(weights.get(kind) or 0) for kind in active_kinds)
    weighted_sum = sum(
        float(normalized.get(kind, {}).get("normalized") or 0) * float(weights.get(kind) or 0)
        for kind in active_kinds
    )

    components: dict[str, int | None] = {
        "physical": None,
        "cognitive": None,
        "time": None,
        "sentiment": None,
        "speech": None,
    }
    for category in components:
        category_kinds = [kind for kind in active_kinds if KIND_CATEGORY[kind] == category]
        category_weight = sum(float(weights.get(kind) or 0) for kind in category_kinds)
        category_weighted_sum = sum(
            float(normalized.get(kind, {}).get("normalized") or 0) * float(weights.get(kind) or 0)
            for kind in category_kinds
        )
        components[category] = (
            round_whole(category_weighted_sum / category_weight) if category_weight > 0 else None
        )

    return {
        "components": components,
        "total": round_whole(weighted_sum / total_weight) if total_weight > 0 else None,
        "breakdown": [
            {
                "kind": kind,
                "raw": normalized.get(kind, {}).get("raw") or 0,
                "normalized": normalized.get(kind, {}).get("normalized") or 0,
                "weight": weights.get(kind) or 0,
                "measured": kind in active_kinds,
                "contribution": (
                    round_tenth(
                        (
                            float(normalized.get(kind, {}).get("normalized") or 0)
                            * float(weights.get(kind) or 0)
                        )
                        / total_weight
                    )
                    if total_weight > 0 and kind in active_kinds
                    else 0
                ),
            }
            # Weighted kinds that were not measured stay listed (measured: false) so stored rows
            # can tell the report what is missing rather than implying a zero.
            for kind in weighted_kinds
        ],
    }

File: apps/web/src/lib/effort-score.ts
import type {
  MeasureCategory,
  MeasureKind,
  Normalization,
  ReportMeasure,
  ReportVideo,
  WeightProfile,
} from "./report-contract";

export type NormalizedKindScore = {
  kind: MeasureKind;
  raw: number;
  normalized: number;
  /** Whether this recording could measure the kind at all. Unmeasured is not zero. */
  measured: boolean;
};

export type NormalizedScores = Partial<Record<MeasureKind, NormalizedKindScore>>;

export type ScoreResult = {
  /** null = nothing in the category was measured for this recording. */
  components: Record<MeasureCategory, number | null>;
  total: number | null;
  breakdown: Array<{
    kind: MeasureKind;
    raw: number;
    normalized: number;
    weight: number;
    measured: boolean;
    contribution: number;
  }>;
};

export const KIND_CATEGORY: Record<MeasureKind, MeasureCategory> = {
  clicks: "physical",
  context_switch: "cognitive",
  keypresses: "physical",
  sentiment: "sentiment",
  scrolls: "physical",
  // Words per minute is speech, not physical effort (preview review B6).
  spoken_word: "speech",
  time_segment: "time",
  utterance: "speech",
};

const SCORE_KIND_ORDER: MeasureKind[] = [
  "clicks",
  "context_switch",
  "keypresses",
  "sentiment",
  "scrolls",
  "spoken_word",
  "time_segment",
  "utterance",
];

// Scale factors are intentionally fixed for TS/Python parity.
// They make the demo fixture land at readable 0-100 values:
// roughly 1.7 switches/min => 63, 18 words/min => 45.
const PER_MINUTE_SCALE: Partial<Record<MeasureKind, number>> = {
  clicks: 1,
  context_switch: 36.75,
  keypresses: 1,
  spoken_word: 2.47,
  utterance: 10,
};

const roundWhole = (value: number) => Math.round(value);
const roundTenth = (value: number) => Math.round(value * 10) / 10;
const clampScore = (value: number) => Math.max(0, Math.min(100, value));

function measuresForKind(measures: ReportMeasure[], kind: MeasureKind) {
  return measures.filter((measure) => measure.kind === kind);
}

function durationMinutes(video: ReportVideo) {
  return video.duration_ms / 60_000;
}

function perMinuteRaw(kind: MeasureKind, measures: ReportMeasure[]) {
  const kindMeasures = measuresForKind(measures, kind);

  if (kind === "keypresses" || kind === "spoken_word") {
    return kindMeasures.reduce((sum, measure) => sum + (measure.value_num ?? 0), 0);
  }

  return kindMeasures.length;
}

function normalizePerMinute(kind: MeasureKind, measures: ReportMeasure[], video: ReportVideo) {
  const raw = perMinuteRaw(kind, measures);
  const scale = PER_MINUTE_SCALE[kind] ?? 1;
  const rate = durationMinutes(video) > 0 ? raw / durationMinutes(video) : 0;

  return {
    raw,
    normalized: roundWhole(clampScore(rate * scale)),
  };
}

/**
 * An utterance counts toward negative narration time at or below this value; the report's
 * "worth attention" line uses the same constant. Keep in sync with the worker's effort_score.py.
 */
export const NEGATIVE_SENTIMENT_THRESHOLD = -0.3

function normalizeNegativeDensity(measures: ReportMeasure[]) {
  const sentimentMeasures = measuresForKind(measures, "sentiment");
  const narrationMs = sentimentMeasures.reduce((sum, measure) => {
    return sum + Math.max(0, (measure.t_end_ms ?? measure.t_start_ms) - measure.t_start_ms);
  }, 0);
  const negativeMs = sentimentMeasures.reduce((sum, measure) => {
    if ((measure.value_num ?? 0) > NEGATIVE_SENTIMENT_THRESHOLD) {
      return sum;
    }

    return sum + Math.max(0, (measure.t_end_ms ?? measure.t_start_ms) - measure.t_start_ms);
  }, 0);
  const share = narrationMs > 0 ? negativeMs / narrationMs : 0;

  return {
    raw: roundTenth(share),
    normalized: roundWhole(clampScore(share * 250)),
  };
}

function normalizeRawMinutes(video: ReportVideo) {
  const minutes = durationMinutes(video);

  return {
    raw: roundTenth(minutes),
    normalized: roundWhole(clampScore(minutes * (100 / 30))),
  };
}

/**
 * Whether this recording could measure the kind at all (mirrors is_measured in the worker).
 * Speech kinds need audio (audio with silence is a real zero); sentiment needs speech to classify;
 * context switches and time come from the video; the experimental physical detectors count only
 * when they produced something.
 */
export function isMeasured(kind: MeasureKind, measures: ReportMeasure[], video: ReportVideo): boolean {
  if (kind === "spoken_word" || kind === "utterance") return video.has_audio !== false;
  if (kind === "sentiment") return measuresForKind(measures, "sentiment").length > 0;
  if (kind === "context_switch" || kind === "time_segment") return true;
  return measuresForKind(measures, kind).length > 0;
}

export function normalize(
  measures: ReportMeasure[],
  video: ReportVideo,
  normalization: Partial<Record<MeasureKind, Normalization>>,
): NormalizedScores {
  return SCORE_KIND_ORDER.reduce<NormalizedScores>((scores, kind) => {
    const mode = normalization[kind];

    if (!mode) {
      return scores;
    }

    const measured = isMeasured(kind, measures, video);

    if (mode === "per_minute") {
      scores[kind] = { kind, ...normalizePerMinute(kind, measures, video), measured };
      return scores;
    }

    if (mode === "neg_density") {
      scores[kind] = { kind, ...normalizeNegativeDensity(measures), measured };
      return scores;
    }

    scores[kind] = { kind, ...normalizeRawMinutes(video), measured };
    return scores;
  }, {});
}

export function score(
  normalized: NormalizedScores,
  weights: WeightProfile["weights"],
): ScoreResult {
  const weightedKinds = SCORE_KIND_ORDER.filter((kind) => (weights[kind] ?? 0) > 0);
  // Only kinds this recording could measure take part, in numerator and denominator alike, so a
  // missing signal never drags the score toward zero. A kind with no normalization is missing.
  const activeKinds = weightedKinds.filter((kind) => normalized[kind]?.measured ?? false);
  const totalWeight = activeKinds.reduce((sum, kind) => sum + (weights[kind] ?? 0), 0);

  const weightedSum = activeKinds.reduce((sum, kind) => {
    return sum + ((normalized[kind]?.normalized ?? 0) * (weights[kind] ?? 0));
  }, 0);

  const components = (["physical", "cognitive", "time", "sentiment", "speech"] as const).reduce(
    (componentScores, category) => {
      const categoryKinds = activeKinds.filter((kind) => KIND_CATEGORY[kind] === category);
      const categoryWeight = categoryKinds.reduce((sum, kind) => sum + (weights[kind] ?? 0), 0);
      const categoryWeightedSum = categoryKinds.reduce((sum, kind) => {
        return sum + ((normalized[kind]?.normalized ?? 0) * (weights[kind] ?? 0));
      }, 0);

      componentScores[category] = categoryWeight > 0
        ? roundWhole(categoryWeightedSum / categoryWeight)
        : null;

      return componentScores;
    },
    {
      physical: null,
      cognitive: null,
      time: null,
      sentiment: null,
      speech: null,
    } as Record<MeasureCategory, number | null>,
  );

  return {
    components,
    total: totalWeight > 0 ? roundWhole(weightedSum / totalWeight) : null,
    // Weighted kinds that were not measured stay listed (measured: false) so the report can say
    // what is missing rather than implying a zero.
    breakdown: weightedKinds.map((kind) => {
      const measured = activeKinds.includes(kind);
      return {
        kind,
        raw: normalized[kind]?.raw ?? 0,
        normalized: normalized[kind]?.normalized ?? 0,
        weight: weights[kind] ?? 0,
        measured,
        contribution: totalWeight > 0 && measured
          ? roundTenth(((normalized[kind]?.normalized ?? 0) * (weights[kind] ?? 0)) / totalWeight)
          : 0,
      };
    }),
  };
}
